File: backend/app/schemas/outbreak_event.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

SeverityLevel = Literal[1, 2, 3]
OutbreakStatus = Literal["pending", "validated", "rejected"]
ValidateAction = Literal["confirm", "correct", "reject"]


class OutbreakEventCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plague: str = Field(..., min_length=2, max_length=50)
    severity: SeverityLevel
    zone_id: int
    model_version: str = Field(default="v0.0", max_length=10)
    source_scan_id: int | None = None

    @field_validator("plague")
    @classmethod
    def normalize_plague(cls, value: str) -> str:
        return value.strip().lower()


class OutbreakEventRead(BaseModel):
    id: int
    plague: str
    severity: int
    zone_id: int
    zone_name: str | None = None
    reported_at: datetime
    model_version: str
    validated: bool
    status: OutbreakStatus = "pending"
    original_plague: str | None = None
    corrected_plague: str | None = None
    display_plague: str
    source_scan_id: int | None = None

    model_config = ConfigDict(from_attributes=True)


class OutbreakEventValidate(BaseModel):
    validated: bool | None = None
    action: ValidateAction | None = None
    corrected_plague: str | None = Field(default=None, max_length=50)
    corrected_severity: SeverityLevel | None = None

    @model_validator(mode="after")
    def normalize_action(self) -> "OutbreakEventValidate":
        if self.action is None and self.validated is not None:
            self.action = "confirm" if self.validated else "reject"
        if self.action is None:
            self.action = "confirm"
        return self

    @field_validator("corrected_plague")
    @classmethod
    def normalize_corrected_plague(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return value.strip().lower()

File: frontend/lib/models/outbreak_event.dart
class OutbreakEvent {
  final int id;
  final String plague;
  final int severity;
  final int zoneId;
  final String? zoneName;
  final DateTime reportedAt;
  final String modelVersion;
  final bool validated;
  final String status;
  final String? originalPlague;
  final String? correctedPlague;
  final String displayPlague;
  final int? sourceScanId;

  OutbreakEvent({
    required this.id,
    required this.plague,
    required this.severity,
    required this.zoneId,
    this.zoneName,
    required this.reportedAt,
    required this.modelVersion,
    required this.validated,
    this.status = "pending",
    this.originalPlague,
    this.correctedPlague,
    required this.displayPlague,
    this.sourceScanId,
  });

  factory OutbreakEvent.fromJson(Map<String, dynamic> json) {
    return OutbreakEvent(
      id: json["id"] as int,
      plague: json["plague"] as String,
      severity: json["severity"] as int,
      zoneId: json["zone_id"] as int,
      zoneName: json["zone_name"] as String?,
      reportedAt: DateTime.parse(json["reported_at"] as String),
      modelVersion: json["model_version"] as String,
      validated: json["validated"] as bool? ?? false,
      status: json["status"] as String? ?? (json["validated"] == true ? "validated" : "pending"),
      originalPlague: json["original_plague"] as String?,
      correctedPlague: json["corrected_plague"] as String?,
      displayPlague: json["display_plague"] as String? ?? json["plague"] as String,
      sourceScanId: json["source_scan_id"] as int?,
    );
  }

  bool get isPending => status == "pending";
  bool get isValidated => status == "validated";
  bool get wasCorrected => correctedPlague != null && correctedPlague!.isNotEmpty;

  String get severityLabel {
    switch (severity) {
      case 1:
        return "Leve";
      case 2:
        return "Moderado";
      case 3:
        return "Alto";
      default:
        return "Desconocido";
    }
  }
}
